docs/architecture/adr-100-data-companion-push-api.md

```diff
@@ -0,0 +1,513 @@
+    // field is optional, depending on the CometBFT node's configuration.
+    optional tendermint.abci.FinalizeBlockResponse finalize_block_response = 2;
+}
+
+// BlockCommittedResponse is either empty upon succes, or returns one or more
+// errors. Note that returning any errors here will cause CometBFT to crash.
+message BlockCommittedResponse {
+    // If an error occurred during the companion's processing of the request.
+    optional Error error = 1;
+}
+
+// Error can be one of several different types of errors.
+message Error {
+    oneof error {
+        UnexpectedBlockError unexpected_block_err = 1;
+        UnexpectedFieldError unexpected_field_err = 2;
+        ExpectedFieldError   expected_field_err   = 3;
+    }
+}
+
+// UnexpectedBlockError is returned by the server when CometBFT sent it a
+// block that is ahead of the block expected by the server.
+message UnexpectedBlockError {
+    // The height of the block expected by the server.
+    int64 expected_height = 1;
+}
+
+// UnexpectedFieldError is returned by the server when CometBFT sent it a
+// message containing an unexpected field. For instance, if the companion
+// expects only the `block` field but not `finalize_block_response`, but somehow
+// the companion receives a value for `finalize_block_response`, it should
+// return this error because it indicates that either CometBFT or the data
+// companion are most likely incorrectly configured.
+message UnexpectedFieldError {
+    // The name of the field whose value was expected to be empty, but was not.
+    string field_name = 1;
+}
+
+// ExpectedFieldError is returned by the server when CometBFT sent it a
+// message containing an empty field value for a field it was expecting to be
+// populated. If this occurs, it indicates that either CometBFT or the data
+// companion are most likely incorrectly configured.
+message ExpectedFieldError {
+    // The name of the field whose value was expected to be populated, but was
+    // empty.
+    string field_name = 1;
+}
+```
+
+### Request Buffering
+
+In order to ensure reliable delivery, while also catering for intermittent
+faults, we should facilitate buffering of data destined for the companion
+service. This data would need to be stored on disk to ensure that CometBFT
+could attempt to resubmit all unsubmitted data upon CometBFT startup.
+
+At present, this data is already stored on disk, and so practically we would
+need to implement some form of background pruning mechanism to remove the data
+we know has been shared with the companion service.
+
+In order to not allow disk storage to grow unbounded, _and_ ensure that we do
+not lose any critical data, we need to impose some kind of limit on the number
+of requests we will buffer before crashing the CometBFT node. In case there is
+manual intervention needed to restore the companion service, CometBFT should
+attempt to flush this buffer upon startup before continuing normal operations.
```

Comment:
Delivering a message to a buffer does not guarantee that it will be received by any specific consumer process. Storing data on disk does not provide any delivery guarantees. If reliable delivery is a requirement, then data cannot be buffered, neither on disk nor in memory; it must be synchronously delivered to its recipients.

Reply:
I'd like to extend my offer again here: https://github.com/cometbft/cometbft/pull/82#discussion_r1157749867